### PR TITLE
(FACT-2028) Add Windows facts

### PR DESCRIPTION
**Pull request by oanatmaria**
Add domain, hostname and fqdn facts for Windows

### CODE REVIEW

File: lib/resolvers/windows/fqdn_resolver.rb [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+          nil
+        end
+
+        def compose_fqdn(hostname, domain)
+          return [hostname, domain].join('.') if hostname && domain
+
+          hostname
+        end
+
+        def build_fact_list(fact_name)
```
Comment (IrimieBogdan) [inline lib/resolvers/windows/fqdn_resolver.rb]: Should return a list of facts

File: lib/resolvers/windows/fqdn_resolver.rb [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+        def retrieve_domain
+          win = Win32Ole.new
+          result = win.exec_query('select DNSDomain from Win32_NetworkAdapterConfiguration where IPEnabled = True')
+          unless result
+            @log.debug 'WMI query returned no results for Win32_NetworkAdapterConfiguration with value DNSDomain.'
+            return
+          end
+          result.each do |network|
+            return network.DNSDomain if network.DNSDomain && !network.DNSDomain.empty?
```
Comment (IrimieBogdan) [inline lib/resolvers/windows/fqdn_resolver.rb]: The condition could be replaced with
```
if network.DNSDomain.present?
```

File: lib/resolvers/windows/fqdn_resolver.rb [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+            subscribe_to_manager
+            result || compose_fqdn(fact_name)
+          end
+        end
+
+        private
+
+        def retrieve_hostname
+          output, _status = Open3.capture2('hostname')
+          output ? output.strip : output
```
Comment (IrimieBogdan) [inline lib/resolvers/windows/fqdn_resolver.rb]: This can be replaced with
```
output&.strip
```

File: lib/resolvers/windows/fqdn_resolver.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+          nil
+        end
+
+        def compose_fqdn(fact_name)
+          domain = retrieve_domain
+          hostname = retrieve_hostname
+          fqdn = nil
+          fqdn = [hostname, domain].join('.') if hostname && domain
+
+          result = { hostname: hostname, domain: domain, fqdn: fqdn || hostname }
```
Comment (sebastian-miclea) [inline lib/resolvers/windows/fqdn_resolver.rb]: Shouldn't this be split in two resolvers hand have the logic that decides which (syscall/ole query) should it use?

File: lib/resolvers/windows/fqdn_resolver.rb (hunk 2 above)
Comment (oanatmaria) [reply lib/resolvers/windows/fqdn_resolver.rb]: undefined method `present?' for nil:NilClass

Comment (IrimieBogdan) [reply lib/resolvers/windows/fqdn_resolver.rb]: present? is only available in rails

**Review by BogdanIrimie — APPROVED** [review]

**Pull request merged**